File: Example/javaapp/src/main/java/com/stripe/example/javaapp/fragment/event/EventAdapter.java
package com.stripe.example.javaapp.fragment.event;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.databinding.BindingAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.stripe.example.javaapp.R;
import com.stripe.example.javaapp.model.Event;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

/**
 * Our [RecyclerView.Adapter] implementation that allows us to update the list of events
 */
public class EventAdapter extends RecyclerView.Adapter<EventHolder> {
    @NotNull private List<Event> events;

    public EventAdapter() {
        super();
        events = new ArrayList<>();
    }

    void updateEvents(@NotNull List<Event> events) {
        this.events = events;
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return events.size();
    }

    @Override
    public void onBindViewHolder(@NotNull EventHolder holder, int position) {
        holder.bind(events.get(position));
    }

    @NotNull
    @Override
    public EventHolder onCreateViewHolder(@NotNull ViewGroup parent, int viewType) {
        return new EventHolder(
                LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_event, parent, false));
    }
}

File: Example/kotlinapp/src/main/java/com/stripe/example/fragment/event/EventAdapter.kt
package com.stripe.example.fragment.event

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.AsyncListDiffer
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.stripe.example.databinding.ListItemEventBinding
import com.stripe.example.model.Event

/**
 * Our [RecyclerView.Adapter] implementation that allows us to update the list of events
 */
class EventAdapter : RecyclerView.Adapter<EventHolder>() {
    private val differ = AsyncListDiffer(this, ItemCallback())

    fun updateEvents(events: List<Event>) {
        differ.submitList(events)
    }

    override fun getItemCount(): Int = differ.currentList.size

    override fun onBindViewHolder(holder: EventHolder, position: Int) {
        holder.bind(differ.currentList[position])
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): EventHolder {
        return EventHolder(
            ListItemEventBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false,
            )
        )
    }

    private class ItemCallback : DiffUtil.ItemCallback<Event>() {
        override fun areItemsTheSame(oldItem: Event, newItem: Event): Boolean {
            return oldItem == newItem
        }

        override fun areContentsTheSame(oldItem: Event, newItem: Event): Boolean {
            return oldItem == newItem
        }
    }
}
